Conch/lib/Conch/Model/DeviceRole.pm

```diff
@@ -0,0 +1,361 @@
+=head1 NAME
+
+Conch::Model::DeviceRole
+
+=head1 DESCRIPTION
+
+A Device Role is a combination of a hardware product and a list of services
+that the box will perform. This identifies the server's "role" in the
+ecosystems.
+
+=cut
+
+package Conch::Model::DeviceRole;
+
+use strict;
+use warnings;
+use utf8;
+use v5.20;
+
+use Moo;
+use Moo::Role::ToJSON;
+use experimental qw(signatures);
+
+use Try::Tiny;
+use Type::Tiny;
+use Types::Standard qw(Str ArrayRef);
+use Types::UUID qw(Uuid);
+
+use List::MoreUtils qw(uniq bremove qsort);
+
+use Conch::Pg;
+
+with("Moo::Role::ToJSON");
+with("Conch::Role::But");
+with("Conch::Role::Timestamps");
+
+
+=head1 ACCESSORS
+
+=over 4
+
+=item id
+
+UUID. Cannot be written by user.
+
+=cut
+
+has 'id' => (
+	is  => 'rwp',
+	isa => Uuid,
+);
+
+
+=item name
+
+String. Required.
+
+=cut
+
+has 'name' => (
+	is       => 'rw',
+	isa      => Str,
+	required => 1,
+);
+
+
+=item hardware_product_id
+
+UUID. Required. FK'd to C<hardware_product(id)>
+
+=cut
+
+has 'hardware_product_id' => (
+	is => 'rw',
+	isa => Uuid,
+	required => 1,
+);
+
+
+=item services
+
+ArrayRef containing the IDs of services bound to this Role
+
+=cut
+
+has 'services' => (
+	is => 'rwp',
+	isa => ArrayRef,
+	default => sub { [] },
+);
+
+
+
+sub _build_serializable_attributes { [qw[
+	id
+	name
+	hardware_product_id
+	services
+	created
+	updated
+	deactivated
+]] }
+
+
+
+
+=back
+
+=head1 METHODS
+
+=head2 from_id
+
+	my $role = Conch::Model::DeviceRole->from_id($uuid);
+
+Returns a DeviceRole given an existing uuid
+
+Includes deactivated objects
+
+=cut
+
+sub from_id ($class, $id) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select r.*, array(
+				select drs.service_id
+				from device_role_services drs
+				where drs.role_id = r.id
+				order by drs.service_id
+			) as services
+			from device_role r
+			where r.id = ?
+		|, $id)->hash;
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->from_id: $_");
+		return undef;
+	};
+
+	return undef unless $ret;
+	return $class->new(_fixup_timestamptzs($ret)->%*);
+}
+
+
+
+=head2 from_name
+
+	my $role = Conch::Model::DeviceRole->from_name($name);
+
+Returns a DeviceRole given an existing name
+
+Includes deactivated objects
```

Comment:
Including deactivated means the query can return 0-or-many results, but this method only returns the first of those. You could end up in a situation where you've deactivated a role, created a new one with the same name, and this method would only return the deactivated role rather than the new one.